plugins/cache/bap_cache.ml

```diff
@@ -0,0 +1,135 @@
+open Core_kernel
+open Regular.Std
+open Bap.Std
+open Bap_cache_types
+
+include Self ()
+
+module Filename = Caml.Filename
+module Utils = Bap_cache_utils
+module Cfg = Bap_cache_config
+
+let (/) = Filename.concat
+
+let cache_dir = Cfg.cache_dir
+let lock_file = "lock"
+
+let read () = Sys.readdir @@ cache_dir ()
+
+let size () =
+  let (+) = Int64.(+) in
+  let dir = cache_dir () in
+  read () |>
+  Array.fold ~init:0L
+    ~f:(fun sz e -> sz + Unix.LargeFile.( (stat @@ dir / e).st_size ))
+
+let with_lock ~f =
+  let lock = cache_dir () / lock_file in
+  let lock = Unix.openfile lock Unix.[O_RDWR; O_CREAT] 0o640 in
+  Unix.lockf lock Unix.F_LOCK 0;
+  protect ~f
+    ~finally:(fun () -> Unix.(lockf lock F_ULOCK 0; close lock))
+
+
+module GC = struct
+
+  let () = Random.self_init ()
+
+  let remove_entry path =
+    try Sys.remove path
+    with exn ->
+      warning "unable to remove entry: %s" (Exn.to_string exn)
+
+  let file_size x = Unix.LargeFile.( (stat x).st_size )
+
+  let remove_entries is_entry min_length size_to_free =
+    let dir = cache_dir () in
+    let rec loop entries freed len =
+      if freed < size_to_free && len > min_length then
```

Comment:
`len > min_length` is a loop invariant, move it away. 

Also, get rid of the fragile `is_entry` instead either populate the array with entries only or separate meta files and data files into separate directories, e.g.,

```
$CACHE_DIR/
 |
 +- config
 +- lock
 +- data
```
This setup will make it easier to clean the cache, by just dropping the `data` folder as well as enable a future upgrade to a tree representation (analogous to the git representation).